File: apps/ios/Sources/Litter/Models/PathDisplay.swift
import Foundation

/// Convert filesystem paths to short, user-facing strings.
///
/// For **local** codex paths, rewrites the app-container Documents home
/// and the app's `NSTemporaryDirectory()` to `~` and `/tmp` so the UI
/// shows `~/projects/foo` and `/tmp/x.txt` instead of
/// `/var/mobile/Containers/Data/Application/<UUID>/Documents/home/codex/...`.
///
/// For **remote** paths, delegates to the existing `abbreviateHomePath`
/// which shortens `/Users/<user>/<subpath>` and `/home/<user>/<subpath>`
/// to `~/<subpath>`.
enum PathDisplay {
    /// Callers pass `isLocal = true` only when `raw` is a path on the
    /// in-process iOS codex. Remote-server paths (SSH/WebSocket) go
    /// through `abbreviateHomePath`.
    static func display(_ raw: String, isLocal: Bool) -> String {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return isLocal ? "~" : trimmed }
        guard isLocal else { return remoteDisplay(trimmed) }
        let home = HomeAnchor.path
        if trimmed == home { return "~" }
        if trimmed.hasPrefix(home + "/") {
            return "~/" + trimmed.dropFirst(home.count + 1)
        }
        let tmp = realTmp()
        if !tmp.isEmpty {
            if trimmed == tmp { return "/tmp" }
            if trimmed.hasPrefix(tmp + "/") {
                return "/tmp/" + trimmed.dropFirst(tmp.count + 1)
            }
        }
        return trimmed
    }

    /// Inverse of `display`. Accepts user-entered display strings (`~/foo`,
    /// `/tmp/x`, or remote `~\foo`) and produces an absolute path for the
    /// selected host.
    static func expand(_ display: String, isLocal: Bool, remoteHome: String? = nil) -> String {
        let trimmed = display.trimmingCharacters(in: .whitespacesAndNewlines)
        guard isLocal else {
            return expandRemoteDisplay(trimmed, remoteHome: remoteHome)
        }
        if trimmed == "~" { return HomeAnchor.path }
        if trimmed.hasPrefix("~/") {
            return HomeAnchor.path + "/" + trimmed.dropFirst(2)
        }
        let tmp = realTmp()
        if !tmp.isEmpty {
            if trimmed == "/tmp" { return tmp }
            if trimmed.hasPrefix("/tmp/") {
                return tmp + "/" + trimmed.dropFirst(5)
            }
        }
        return trimmed
    }

    private static func realTmp() -> String {
        let raw = NSTemporaryDirectory()
        // Strip trailing slash so comparisons are uniform with prefix
        // matching that adds `/`.
        if raw.hasSuffix("/") { return String(raw.dropLast()) }
        return raw
    }

    private static func remoteDisplay(_ trimmed: String) -> String {
        let remote = RemotePath.parse(path: trimmed)
        guard remote.isWindows() else {
            return abbreviateHomePath(trimmed)
        }
        return abbreviateWindowsHome(remote) ?? remote.asString()
    }

    private static func expandRemoteDisplay(_ display: String, remoteHome: String?) -> String {
        guard let home = remoteHome?.trimmingCharacters(in: .whitespacesAndNewlines),
              !home.isEmpty else {
            return display
        }
        let remoteHomePath = RemotePath.parse(path: home)
        let normalizedHome = remoteHomePath.asString()
        if remoteHomePath.isWindows() {
            guard display == "~" || display.hasPrefix("~\\") || display.hasPrefix("~/") else {
                return display
            }
            if display == "~" { return normalizedHome }
            let suffix = String(display.dropFirst(2)).replacingOccurrences(of: "/", with: "\\")
            return appendWindowsSuffix(suffix, to: normalizedHome)
        }
        guard display == "~" || display.hasPrefix("~/") else {
            return display
        }
        if display == "~" { return normalizedHome }
        let suffix = String(display.dropFirst(2))
        return normalizedHome.hasSuffix("/") ? normalizedHome + suffix : normalizedHome + "/" + suffix
    }

    private static func appendWindowsSuffix(_ suffix: String, to home: String) -> String {
        let trimmedHome = home.hasSuffix("\\") ? String(home.dropLast()) : home
        guard !suffix.isEmpty else { return trimmedHome }
        return trimmedHome + "\\" + suffix
    }

    private static func abbreviateWindowsHome(_ remote: RemotePath) -> String? {
        let segments = remote.segments()
        guard segments.count >= 3 else { return nil }
        guard segments[1].label.caseInsensitiveCompare("Users") == .orderedSame else {
            return nil
        }
        let remainder = segments.dropFirst(3).map(\.label)
        guard !remainder.isEmpty else { return "~" }
        return "~\\" + remainder.joined(separator: "\\")
    }
}

File: apps/android/app/src/main/java/com/litter/android/state/PathDisplay.kt
package com.litter.android.state

import android.content.Context

/**
 * Convert filesystem paths to short, user-facing strings.
 *
 * For **local** codex paths, rewrites the `HomeAnchor.path(context)` and
 * `$TMPDIR` to `~` and `/tmp` so the UI shows `~/projects/foo` and
 * `/tmp/x.txt` instead of the raw `/data/user/0/com.sigkitten.litter/files/...`
 * absolute paths.
 *
 * For **remote** paths, shortens paths under a resolved remote home directory
 * when the caller provides one.
 */
object PathDisplay {
    /**
     * Callers pass [isLocal] `= true` only when [raw] is a path on the
     * in-process Android codex. Remote-server paths can be abbreviated when
     * [remoteHome] is known.
     */
    fun display(
        raw: String,
        isLocal: Boolean,
        context: Context,
        remoteHome: String? = null,
    ): String {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return if (isLocal) "~" else trimmed
        if (!isLocal) return remoteDisplay(trimmed, remoteHome)
        val home = HomeAnchor.path(context)
        if (trimmed == home) return "~"
        if (trimmed.startsWith("$home/")) return "~/" + trimmed.substring(home.length + 1)
        val tmp = realTmp()
        if (tmp.isNotEmpty()) {
            if (trimmed == tmp) return "/tmp"
            if (trimmed.startsWith("$tmp/")) return "/tmp/" + trimmed.substring(tmp.length + 1)
        }
        return trimmed
    }

    /** Inverse of [display] for user-entered display strings on the selected server. */
    fun expand(
        display: String,
        isLocal: Boolean,
        context: Context,
        remoteHome: String? = null,
    ): String {
        val trimmed = display.trim()
        if (!isLocal) return expandRemoteDisplay(trimmed, remoteHome)
        if (trimmed == "~") return HomeAnchor.path(context)
        if (trimmed.startsWith("~/")) return HomeAnchor.path(context) + "/" + trimmed.substring(2)
        val tmp = realTmp()
        if (tmp.isNotEmpty()) {
            if (trimmed == "/tmp") return tmp
            if (trimmed.startsWith("/tmp/")) return "$tmp/" + trimmed.substring(5)
        }
        return trimmed
    }

    private fun realTmp(): String {
        // Set by `Java_com_litter_android_core_bridge_UniffiInit_nativeBridgeInit`
        // at JNI boot. Strip trailing slash so comparisons are uniform.
        val raw = System.getenv("TMPDIR") ?: return ""
        return if (raw.endsWith("/")) raw.dropLast(1) else raw
    }

    private fun remoteDisplay(raw: String, remoteHome: String?): String {
        val home = remoteHome?.trim().orEmpty()
        if (home.isEmpty()) return raw
        val windows = isWindowsPath(home)
        val normalizedRaw = if (windows) raw.replace('/', '\\') else raw
        val normalizedHome = if (windows) home.replace('/', '\\').trimEnd('\\') else home.trimEnd('/')
        if (normalizedHome.isEmpty()) return raw
        if (windows) {
            if (normalizedRaw.equals(normalizedHome, ignoreCase = true)) return "~"
            val prefix = "$normalizedHome\\"
            if (normalizedRaw.startsWith(prefix, ignoreCase = true)) {
                return "~\\" + normalizedRaw.substring(prefix.length)
            }
            return raw
        }
        if (normalizedRaw == normalizedHome) return "~"
        val prefix = "$normalizedHome/"
        if (normalizedRaw.startsWith(prefix)) return "~/" + normalizedRaw.substring(prefix.length)
        return raw
    }

    private fun expandRemoteDisplay(display: String, remoteHome: String?): String {
        val home = remoteHome?.trim().orEmpty()
        if (home.isEmpty()) return display
        val windows = isWindowsPath(home)
        val normalizedHome = if (windows) home.replace('/', '\\').trimEnd('\\') else home.trimEnd('/')
        if (normalizedHome.isEmpty()) return display
        if (windows) {
            if (display == "~") return normalizedHome
            if (display.startsWith("~\\") || display.startsWith("~/")) {
                return normalizedHome + "\\" + display.substring(2).replace('/', '\\')
            }
            return display
        }
        if (display == "~") return normalizedHome
        if (display.startsWith("~/")) return "$normalizedHome/${display.substring(2)}"
        return display
    }

    private fun isWindowsPath(path: String): Boolean =
        path.length >= 2 && path[0].isLetter() && path[1] == ':'
}
